pkg/duck/sinks.go

```diff
@@ -17,50 +17,84 @@ limitations under the License.
+// SinkReconciler is a helper for Sources. It triggers
+// reconciliation on creation, updates to or deletion of the source's sink.
+type SinkReconciler struct {
+	tracker             tracker.Interface
+	sinkInformerFactory pkgapisduck.InformerFactory
+}
+
+// NewSinkReconciler creates and initializes a new SinkReconciler
+func NewSinkReconciler(opt reconciler.Options, callback func(string)) *SinkReconciler {
+	ret := &SinkReconciler{}
+
+	ret.tracker = tracker.New(callback, opt.GetTrackerLease())
+	ret.sinkInformerFactory = &pkgapisduck.CachedInformerFactory{
+		Delegate: &pkgapisduck.EnqueueInformerFactory{
+			Delegate: &pkgapisduck.TypedInformerFactory{
+				Client:       opt.DynamicClientSet,
+				Type:         &duckv1alpha1.AddressableType{},
+				ResyncPeriod: opt.ResyncPeriod,
+				StopChannel:  opt.StopChannel,
+			},
+			EventHandler: reconciler.Handler(ret.tracker.OnChanged),
+		},
 	}
 
-	rc := dynamicClient.Resource(duckapis.KindToResource(sink.GroupVersionKind()))
-	if rc == nil {
-		return "", fmt.Errorf("failed to create dynamic client resource")
+	return ret
+}
+
+// GetSinkURI registers the given object reference with the tracker and if possible,
+// retrieves the sink URI
+func (r *SinkReconciler) GetSinkURI(sinkObjRef *corev1.ObjectReference, source interface{}, sourceDesc string) (string, error) {
+	if sinkObjRef == nil {
+		return "", fmt.Errorf("sink ref is nil")
 	}
 
-	u, err := rc.Namespace(namespace).Get(sink.Name, metav1.GetOptions{})
-	if err != nil {
-		return "", err
+	if err := r.tracker.Track(*sinkObjRef, source); err != nil {
+		return "", fmt.Errorf("Error tracking sink '%+v' for source %q: %+v", sinkObjRef, sourceDesc, err)
 	}
 
-	objIdentifier := fmt.Sprintf("\"%s/%s\" (%s)", u.GetNamespace(), u.GetName(), u.GroupVersionKind())
+	// K8s Services are special cased. They can be called, even though they do not satisfy the
+	// Callable interface.
+	if sinkObjRef.APIVersion == "v1" && sinkObjRef.Kind == "Service" {
+		return DomainToURL(names.ServiceHostName(sinkObjRef.Name, sinkObjRef.Namespace)), nil
+	}
 
-	t := duckv1alpha1.AddressableType{}
-	err = duck.FromUnstructured(u, &t)
+	gvr, _ := meta.UnsafeGuessKindToResource(sinkObjRef.GroupVersionKind())
+	_, lister, err := r.sinkInformerFactory.Get(gvr)
 	if err != nil {
-		return "", fmt.Errorf("failed to deserialize sink %s: %v", objIdentifier, err)
+		return "", fmt.Errorf("Error getting a lister for a sink resource '%+v': %+v", gvr, err)
 	}
 
-	if t.Status.Address == nil {
-		return "", fmt.Errorf("sink %s does not contain address", objIdentifier)
+	sinkObj, err := lister.ByNamespace(sinkObjRef.Namespace).Get(sinkObjRef.Name)
+	if err != nil {
+		return "", fmt.Errorf("Error fetching sink %+v for source %q: %v", sinkObjRef, sourceDesc, err)
 	}
 
-	if t.Status.Address.Hostname == "" {
-		return "", fmt.Errorf("sink %s contains an empty hostname", objIdentifier)
+	sink, ok := sinkObj.(*duckv1alpha1.AddressableType)
```

Comment:
I think there might be a helper method on AddressableType now that helps clean up trying to figure out its address.

Reply:
Hm, didn't find in https://github.com/knative/pkg/blob/master/apis/duck ...